include/boost/hana/index_if.hpp

```diff
@@ -0,0 +1,116 @@
+#include <boost/hana/concept/foldable.hpp>
+#include <boost/hana/concept/iterable.hpp>
+#include <boost/hana/config.hpp>
+#include <boost/hana/detail/decay.hpp>
+#include <boost/hana/detail/index_if.hpp>
+#include <boost/hana/fwd/basic_tuple.hpp>
+#include <boost/hana/integral_constant.hpp>
+#include <boost/hana/length.hpp>
+#include <boost/hana/optional.hpp>
+
+#include <cstddef>
+#include <limits>
+#include <utility>
+
+
+BOOST_HANA_NAMESPACE_BEGIN
+    //! @cond
+    template <typename Xs, typename Pred>
+    constexpr auto index_if_t::operator()(Xs&& xs, Pred&& pred) const {
+        using S = typename hana::tag_of<Xs>::type;
+        using IndexIf = BOOST_HANA_DISPATCH_IF(index_if_impl<S>,
+            hana::Iterable<S>::value
+        );
+
+    #ifndef BOOST_HANA_CONFIG_DISABLE_CONCEPT_CHECKS
+        static_assert(hana::Iterable<S>::value,
+        "hana::index_if(xs, pred) requires 'xs' to be a Iterable");
+    #endif
+
+        return IndexIf::apply(static_cast<Xs&&>(xs), static_cast<Pred&&>(pred));
+    }
+    //! @endcond
+
+    namespace detail {
+        template <std::size_t i, std::size_t N, bool Done>
+        struct iterate_while;
+
+        template <std::size_t i, std::size_t N>
+        struct iterate_while<i, N, false> {
+            template <typename F, typename Pred>
+            using f = typename iterate_while<i + 1, N,
+                static_cast<bool>(detail::decay<decltype(
+                    std::declval<Pred>()(std::declval<F>()(hana::size_c<i>)))>::type::value)
+            >::template f<F, Pred>;
+        };
+
+        template <std::size_t N>
+        struct iterate_while<N, N, false> {
+            template <typename F, typename Pred>
+            using f = hana::optional<>;
+        };
+
+        template <std::size_t i, std::size_t N>
+        struct iterate_while<i, N, true> {
+            template <typename F, typename Pred>
+            using f = hana::optional<hana::size_t<i - 1>>;
+        };
+
+        template <typename Xs>
+        struct partial_at {
+            Xs const& xs;
+
+            template <typename I>
+            constexpr decltype(auto) operator()(I i) const {
+                return hana::at(xs, i);
+            }
+        };
+    }
+
+    template <typename Tag>
+    struct index_if_impl<Tag, when<Foldable<Tag>::value>> {
+        template <typename Xs, typename Pred>
+        static constexpr auto apply(Xs const& xs, Pred const&)
+            -> typename detail::iterate_while<0,
+                decltype(hana::length(xs))::value, false>
+                    ::template f<detail::partial_at<Xs>, Pred>
+        { return {}; }
+    };
+
+    template <typename It>
+    struct index_if_impl<It, when<!Foldable<It>::value>> {
+        template <typename Xs, typename Pred>
+        static constexpr auto apply(Xs const&, Pred const&)
+            -> typename detail::iterate_while<hana::size_c<0>,
+                std::numeric_limits<std::size_t>::max(), false>
```

Comment:
Since `size_t` is unsigned, and unsigned underflow is defined to wrap, I would suggest being bold and just using `static_cast<size_t>(-1)`. I normally wouldn't do this, but this will allow us to get rid of the `<limits>` include.